### PR TITLE
[OUR415-313] Converts search results header to a layout component with composable children

### PR DESCRIPTION
**Pull request by rosschapman**
🎟️ https://www.notion.so/exygy/Browse-experience-distinction-bugs-12e3433f03d080cabd60e6473d97b2a8?pvs=4

This changeset addresses the AC in the ticket `The browse experience for services should not display the “Clear Search” button next to the # of results (that is only reserved for the search experience)`. 

---

My first thought was to add a prop to the `SearchResults` component for `showClearSearchButton`. But it then dawned on me that there's really no reason for the `SearchResults` component to exist as a somewhat inflexible layout component. Instead, I made its existing child components easier for reuse and composition.

There is some redundancy of code now across Browse and Search pages. Like handlers for map actions and first result focus, etc... I'm still thinking about how to consolidate these. But I don't think it's a huge deal that there is just some x2 repetition at the moment. It frees us up to think about these experiences very separately while we're still in early stages. 

- Removes `SearchResults` component
- Creates a `SearchResultsHeader` component with composable children
- Creates a `NoSearchResultsDisplay` component
- Updates the `SearchResultsPage` and `SearchDiscoveryPage` components to use these new shared components
- Adds a component test to check for the existence of the Clear Search button. This new test will hopefully serve as a foundation for additional tests of the results pages!

### CODE REVIEW

File: app/components/search/SearchResults/SearchResults.tsx [hunk 1]
```diff
@@ -12,7 +12,7 @@ import {
 } from "react-instantsearch";
 import styles from "./SearchResults.module.scss";
 import ClearSearchButton from "../Refinements/ClearSearchButton";
-import { Loader } from "components/ui";
+import { Loader } from "components/ui/Loader";
```
Comment (rosschapman) [inline app/components/search/SearchResults/SearchResults.tsx:15]: **note:** See https://github.com/Exygy/askdarcel-web/pull/264

File: app/components/ui/NoSearchResultsDisplay.tsx [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import styles from "components/search/SearchResults/SearchResults.module.scss";
+import ClearSearchButton from "components/search/Refinements/ClearSearchButton";
+
+export const NoSearchResultsDisplay = ({ query }: { query: string }) => (
+  <div className={`${styles.noResultsMessage}`}>
+    <div className={styles.noResultsText}>
+      No results {query && `for ${` "${query}" `}`} found in your area.
+      <br /> Try a different location, filter, or search term.
+    </div>
+    {query && <ClearSearchButton />}
```
Comment (rosschapman) [inline app/components/ui/NoSearchResultsDisplay.tsx:11]: **note:** While this component is used for the Browse Experience, theoretically there will never be a query in that case so the Clear Search button will not show.

File: jest.config.ts [hunk 3]
```diff
@@ -67,7 +67,9 @@ const config: Config = {
   // globalTeardown: undefined,
 
   // A set of global variables that need to be available in all test environments
-  // globals: {},
+  globals: {
+    CONFIG: {},
+  },
```
Comment (rosschapman) [inline jest.config.ts:72]: **note:** Mocks config vars in the tests. We can always add useful mocks here later as needed.

File: jest.config.ts [hunk 4]
```diff
@@ -91,6 +93,9 @@ const config: Config = {
   moduleNameMapper: {
     // Ensures style imports don't break tests
     "\\.(css|scss)$": "<rootDir>/test/jest/__mocks__/styleMock.ts",
+    "react-markdown": "<rootDir>/test/jest/__mocks__/react-markdown.tsx",
+    "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$":
+      "<rootDir>/test/jest/__mocks__/fileMock.ts",
```
Comment (rosschapman) [inline jest.config.ts:98]: **note:** More mocks for imported node modules that aren't transpiled and other file types. Jest isn't running Webpack so it doesn't know what to do when it encounters weird code.

File: test/jest/__mocks__/fileMock.ts [hunk 5]
```diff
@@ -0,0 +1,4 @@
+// Webpack is not available in tests to handle file imports. Our test config
+// tells jest to replace any file imports it sees with this empty string.
+const str = "";
+export default str;
```
Comment (rosschapman) [inline test/jest/__mocks__/fileMock.ts:4]: **note:** Apparently it's a lint rule to not export a primitive directly. 🤷🏻

File: app/pages/ServiceDiscoveryResults/ServiceDiscoveryResults.tsx [hunk 6]
```diff
@@ -1,21 +1,35 @@
-import React, { useState, useEffect } from "react";
```
Comment (adriencyberspace) [inline app/pages/ServiceDiscoveryResults/ServiceDiscoveryResults.tsx:1]: **suggestion (non-blocking):** What if we renamed these components to something like SearchExperience and BrowseExperience? It's getting hard to keep track of !

File: app/pages/SearchResultsPage/SearchResultsPage.test.tsx [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import React from "react";
+import { InstantSearch } from "react-instantsearch-core";
+import { render, screen, waitFor } from "@testing-library/react";
+import SearchResults from "components/search/SearchResults/SearchResults";
```
Comment (adriencyberspace) [inline app/pages/SearchResultsPage/SearchResultsPage.test.tsx:4]: **chore:** rm SearchResults

**Review by adriencyberspace — APPROVED** [review]
Couple small chores and please make a ticket for the `transform` issue! But LGTM

File: app/components/search/Refinements/BrowseRefinementList.tsx [hunk 8]
```diff
@@ -43,9 +43,11 @@ const BrowseRefinementList = ({ attribute, transform }: Props) => {
     setChecked(checked);
   };
 
+  const transformedItems = transform === undefined ? items : transform(items);
```
Comment (rosschapman) [inline app/components/search/Refinements/BrowseRefinementList.tsx:46]: **note:** @adriencyberspace ultimately I moved this transformation execution out of the algolia hook to fix the performance bottleneck and give our code . This works, the infinite looping is gone.

File: app/pages/ServiceDiscoveryResults/ServiceDiscoveryResults.tsx (hunk 6 above)
Comment (rosschapman) [reply app/pages/ServiceDiscoveryResults/ServiceDiscoveryResults.tsx:1]: Yes we need to do that, I thought the same thing.

File: app/pages/ServiceDiscoveryResults/ServiceDiscoveryResults.tsx [hunk 9]
```diff
@@ -1,21 +1,31 @@
...
+import { SearchResult } from "components/search/SearchResults/SearchResult";
+import {
+  TransformedSearchHit,
+  transformSearchResults,
+} from "models/SearchHits";
+import { useInstantSearch, usePagination } from "react-instantsearch";
+import ResultsPagination from "components/search/Pagination/ResultsPagination";
+import searchResultsStyles from "components/search/SearchResults/SearchResults.module.scss";
+import { NoSearchResultsDisplay } from "components/ui/NoSearchResultsDisplay";
+import { SearchResultsHeader } from "components/ui/SearchResultsHeader";
```
[the remark this answers is not in the document]
Comment (rosschapman) [reply app/pages/ServiceDiscoveryResults/ServiceDiscoveryResults.tsx:27]: It's being used now

**Pull request merged**